### PR TITLE
Fixes #8127: On agent, check inventory before sending it to Rudder server

### PR DESCRIPTION
**Pull request by peckpeck**
https://www.rudder-project.org/redmine/issues/8127

### CODE REVIEW

File: initial-promises/node-server/inventory/1.0/fusionAgent.cf [hunk 1]
```diff
@@ -451,6 +453,24 @@ bundle agent addInformationsToInventory
         edit_defaults => def_no_backup;
 }
 
+# Check that te inventory file is valid
```
Comment (ncharles) [inline initial-promises/node-server/inventory/1.0/fusionAgent.cf]: typo: the inventory

File: initial-promises/node-server/inventory/1.0/fusionAgent.cf [hunk 2]
```diff
@@ -451,6 +453,24 @@ bundle agent addInformationsToInventory
...
+bundle agent checkInventoryFile
+{
+  vars:
+      "inventory_file" slist => findfiles("${g.rudder_var_tmp}/inventory/*.ocs");
+  classes:
+      "file_exist" expression => filesexist("@{inventory_file}");
+    file_exist::
```
Comment (ncharles) [inline initial-promises/node-server/inventory/1.0/fusionAgent.cf]: i hate cf_null

File: initial-promises/node-server/inventory/1.0/fusionAgent.cf [hunk 3]
```diff
@@ -451,6 +453,24 @@ bundle agent addInformationsToInventory
...
+      "inventory_valid" expression => returnszero("${this.promise_dirname}/test-inventory.pl ${inventory_file}"),
+                        scope => "namespace";
+
+  reports:
+    file_present.!inventory_valid::
```
Comment (ncharles) [inline initial-promises/node-server/inventory/1.0/fusionAgent.cf]: don't you mean file_exist ?

File: initial-promises/node-server/inventory/1.0/fusionAgent.cf [hunk 4]
```diff
@@ -451,6 +453,24 @@ bundle agent addInformationsToInventory
         edit_defaults => def_no_backup;
 }
 
+# Check that te inventory file is valid
+bundle agent checkInventoryFile
+{
+  vars:
+      "inventory_file" slist => findfiles("${g.rudder_var_tmp}/inventory/*.ocs");
+  classes:
+      "file_exist" expression => filesexist("@{inventory_file}");
```
Comment (ncharles) [inline initial-promises/node-server/inventory/1.0/fusionAgent.cf]: could you rename this to "inventory_file_exists" ?

File: initial-promises/node-server/inventory/1.0/fusionAgent.cf [hunk 5]
```diff
@@ -451,6 +453,24 @@ bundle agent addInformationsToInventory
...
+    file_exist::
+      "inventory_valid" expression => returnszero("${this.promise_dirname}/test-inventory.pl ${inventory_file}"),
+                        scope => "namespace";
+
+  reports:
+    file_present.!inventory_valid::
+      "@@Inventory@@result_error@@inventory-all@@inventory-all@@00@@inventory@@None@@${g.execRun}##${g.uuid}@#Generated inventory has been detected as invalid";
+      # Print information to the user in addition to the reporting
+      "WARNING: your inventory is not valid, not sending it!";
```
Comment (ncharles) [inline initial-promises/node-server/inventory/1.0/fusionAgent.cf]: I think it should display in a more visible way like what we do for the update error
**********************************************************
* Generated inventory is not valid, as it is missing
* mandatory fields.
* Not sending it to the Rudder Server
* You can check the invalid inventory in ${g.rudder_var_tmp}/inventory/
*********************************************************************************************

or something like that
If it were sent at the end of the run, it is even better

File: initial-promises/node-server/inventory/1.0/test-inventory.pl [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+# Test OPERATINGSYSTEM/KERNEL_VERSION or if OPERATINGSYSTEM/KERNEL_NAME.toLowerCase == aix: HARDWARE/OSVERSION
+if(defined $os->{KERNEL_NAME} && lc($os->{KERNEL_NAME}) eq "aix") {
+  if(!defined $tree->{REQUEST}->{CONTENT}->{HARDWARE}->{OSVERSION}) {
+    print "ERROR: No HARDWARE/OSVERSION\n";
+    exit 1;
+  }
+} else {
+  if(!defined $os->{KERNEL_VERSION}) {
+    print "ERROR: No OPERATINGSYSTEM/KERNEL_VERSION\n";
```
Comment (ncharles) [inline initial-promises/node-server/inventory/1.0/test-inventory.pl]: error messages could be a bit more helpful, like 
ERROR: No valid OPERATINGSYSTEM/KERNEL_VERSION entry in the inventory

File: initial-promises/node-server/inventory/1.0/test-inventory.pl [hunk 7]
```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/perl
```
Comment (ncharles) [inline initial-promises/node-server/inventory/1.0/test-inventory.pl]: I think this should be in tools folder if possible

Comment (ncharles) [inline initial-promises/node-server/inventory/1.0/test-inventory.pl]: If it is not in the tools folder, then it should also be on the Technique folder, and you'll need to escape the & in the .st file

Comment (ncharles) [inline initial-promises/node-server/inventory/1.0/test-inventory.pl]: Will this script work on Windows, Solaris and AIX ? (out of the box I mean)

File: initial-promises/node-server/inventory/1.0/fusionAgent.cf (hunk 2 above)
Comment (peckpeck) [inline initial-promises/node-server/inventory/1.0/fusionAgent.cf]: me too

**Comment by jooooooon** [comment]
Looking good otherwise! 👍

File: initial-promises/node-server/inventory/1.0/test-inventory.pl (hunk 6 above)
Comment (Kegeruneku) [inline initial-promises/node-server/inventory/1.0/test-inventory.pl]: Agreed

**Comment by peckpeck** [comment]
Commit modified

**Comment by peckpeck** [comment]
Commit modified

**Comment by ncharles** [comment]
This looks correct to me. You need to rebase it, and validation from someone fluent in perl (like @Kegeruneku )

**Comment by peckpeck** [comment]
PR rebased

**Comment by peckpeck** [comment]
Commit modified

**Comment by peckpeck** [comment]
OK, merging this PR

**Pull request merged**